### PR TITLE
Update PDF annotations when changing bookmark text

### PR DESCRIPTION
**Pull request by Toromtomtom**
This changes the PDF annotation contents each time the corresponding bookmark text is edited. When the annotations are written to the PDF file, the annotation contents are saved as well into the PDF.

In my initial tests it works - I can see the bookmark texts in the PDF viewer Okular. However, there are still some rough edges (partly due to my inexperience with the code base and with Lua):
- To touch as little existing code as possible, I duplicated some code. Once verified that everything works, some refactoring should be done.
- More error checking/reporting should be added.
- Tests should be added for the feature.

Depends on some [low-level additions to koreader-base](https://github.com/koreader/koreader-base/pull/1327).

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7411)

### CODE REVIEW

File: frontend/apps/reader/modules/readerbookmark.lua [hunk 1]
```diff
@@ -597,9 +597,11 @@ end
 function ReaderBookmark:renameBookmark(item, from_highlight)
     if from_highlight then
         -- Called by ReaderHighlight:editHighlight, we need to find the bookmark
+        local pboxes = item.pboxes
         for i=1, #self.bookmarks do
             if item.datetime == self.bookmarks[i].datetime and item.page == self.bookmarks[i].page then
                 item = self.bookmarks[i]
+                item.pboxes = pboxes
```
Comment (Toromtomtom) [inline frontend/apps/reader/modules/readerbookmark.lua:604]: We need to save the pboxes from the (highlight) item, as we need them later to identify the PDF annotation belonging to this bookmark.

**Comment by poire-z** [comment]
I'm totally unfamiliar with PDF, annotations, highlight in PDF (and if these are different things) and how people use the "Save document" feature.
But I believe currently, any stuff that is to be written in the PDF gets handled at document quit time: that's where we get the prompt to write stuff into the PDF - so I guess that's also when it happens when people have disable prompt to use "always".
Was worried you would write stuff to disk on any wrong highlighy I could make, but I see that you just somehow cloned PdfDocument:saveHighlight() to do not much more than what it does :)
So, how does that work ? These 2 functions just add some stuff in memory when called (so, no I/O write to disk ?), and it's only at close time that things get flushed to disk by PdfDocument:writeDocument() ?

**Comment by Toromtomtom** [comment]
> But I believe currently, any stuff that is to be written in the PDF gets handled at document quit time: that's where we get the prompt to write stuff into the PDF - so I guess that's also when it happens when people have disable prompt to use "always".
> Was worried you would write stuff to disk on any wrong highlighy I could make, but I see that you just somehow cloned PdfDocument:saveHighlight() to do not much more than what it does :)
> So, how does that work ? These 2 functions just add some stuff in memory when called (so, no I/O write to disk ?), and it's only at close time that things get flushed to disk by PdfDocument:writeDocument() ?

Yes, that seems to be the case. For example, I observed that, when I added/edited an annotation but killed the emulator before it wrote the file, the PDF file was unchanged.

The mupdf interface works on the in-memory representation of the PDF file. Interestingly, that in-memory representation also contains PDF annotations for which no highlight/bookmark in koreader exists. Those can in principle be edited as well, although that is quite confusing UI-wise. I therefore made sure that editing PDF annotations is only possible if a corresponding koreader highlight/bookmark exists.

**Review by Frenzie — APPROVED** [review]
I'm not really familiar with the API but it looks okay to me.

**Comment by Toromtomtom** [comment]
I will try to do some further testing later today and polish this a bit. Then it will be ready for merging, I guess.

**Comment by Toromtomtom** [comment]
The last commit fixed a crash when editing an annotation in a non-PDF document. From my point of view this is now ready to go.

I will try to do some refactoring and submit a new PR soon.

**Comment by Toromtomtom** [comment]
Anything I can do to move this forward? It seems like this PR has already been approved, so I'm hesitant to make substantial changes now. Or shall I continue pushing changes? There is still some refactoring to be done.

**Comment by Frenzie** [comment]
For me it doesn't really matter in this instance. I can merge it like this and then you can open a follow-up PR, or you could continue adding more commits to this PR.

**Comment by Toromtomtom** [comment]
OK, this should now be good to go. 

I also added a function that deletes annotations from the in-memory representation of the PDF file when the corresponding highlight is removed. Before this, annotations that were added and then removed in koreader were being written to the PDF.

File: frontend/document/pdfdocument.lua [hunk 2]
```diff
@@ -217,6 +194,66 @@ function PdfDocument:saveHighlight(pageno, item)
...
+    local can_write = self:_checkIfWritable()
+    if can_write ~= true then return can_write end
+
+    self.is_edited = true
+    local quadpoints, n = self:_quadpointsFromPboxes(item.pboxes)
+    local page = self._document:openPage(pageno)
+    local annot = page:getMarkupAnnotation(quadpoints, n)
+    if annot ~= nil then
+        page:deleteMarkupAnnotation(annot)
+    end
```
Comment (NiLuJe) [inline frontend/document/pdfdocument.lua:207]: `page` isn't `close`d here.

**Review by NiLuJe — CHANGES_REQUESTED** [review]
Besides the `page` mishap, looks good, thanks!

**Review by NiLuJe — APPROVED** [review]
:lgtm:

**Pull request merged**